packages/ag-charts-website/src/components/tabs/TabsWithHtmlChildren.astro

```diff
@@ -9,6 +10,110 @@ const tabsData = extractTabsData(html);
 const { tabItemIdPrefix } = Astro.props;
 ---
 
-<TabsWithHtmlChildren client:load tabsData={tabsData} tabItemIdPrefix={tabItemIdPrefix}>
-    <Fragment set:html={html} />
-</TabsWithHtmlChildren>
+<tabs-with-html-children data-tabs-data={JSON.stringify(tabsData)} data-tab-Item-id-prefix={tabItemIdPrefix}>
+    <div class:list={['tabs-outer', styles.tabsOuter]}>
+        <header class={'tabs-header'}>
+            <ul class="tabs-nav-list" role="tablist">
+                {
+                    tabsData.map(({ id, label }) => {
+                        const tabId = getTabId({ id: id!, prefix: tabItemIdPrefix });
+                        return (
+                            <li id={tabId} class="tabs-nav-item" role="presentation">
+                                <button class:list={['button-style-none', 'tabs-nav-link']} role="tab">
+                                    {label}
+                                </button>
+                            </li>
+                        );
+                    })
+                }
+            </ul>
+        </header>
+        <div class="tabs-content" role="tabpanel">
+            <Fragment set:html={html} />
+        </div>
+    </div>
+</tabs-with-html-children>
+
+<script>
+    import { TAB_ID_PROP } from './constants';
+    import type { TabData } from './types';
+    import { getTabId } from './utils';
+    import { navigate, browserHistory } from '@utils/navigation';
+
+    const TAB_BUTTON_SELECTOR = '.tabs-nav-item button';
+    const TABS_CONTENT_SELECTOR = '.tabs-content';
+
+    class TabsWithHtmlChildren extends HTMLElement {
+        tabsData: TabData[];
+        selectedIndex: number;
+        tabItemIdPrefix: string;
+
+        constructor() {
+            super();
+
+            this.tabItemIdPrefix = this.dataset.tabItemIdPrefix || '';
+            this.tabsData = this.dataset.tabsData ? JSON.parse(this.dataset.tabsData) : [];
+
+            const hashTabIndex = this.getTabIndexFromHash(window.location?.hash);
+            this.selectedIndex = hashTabIndex >= 0 ? hashTabIndex : 0;
+
+            this.initNavButtonHandlers();
+            this.updateTabs();
+            this.listenToHashChanges();
+        }
+
+        get selected(): TabData {
+            return this.tabsData[this.selectedIndex];
+        }
+
+        listenToHashChanges() {
+            browserHistory?.listen(({ location }) => {
```

Comment:
should we "unlisten" when the component is removed, or is that not a real use case?

Reply:
I figured the site would be loading new pages since it's more of a Multipage app with Astro, so would clean up naturally. However, it's a good shout 👍🏼 , and I should really be using `connectedCallback` and `disconnectedCallback` for this.